pkg/utils/authcheck/authcheck.go

```diff
@@ -0,0 +1,132 @@
+	// corresponding k8s service account and google service account establish a correct relationship.
+	resource = "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token"
+	// scope is used as the scope to get token from default credential.
+	scope = "https://www.googleapis.com/auth/cloud-platform"
+	// authMessage is the key words to determine if a termination log is about authentication.
+	authMessage = "checking authentication"
+)
+
+type AuthenticationCheck interface {
+	Check(ctx context.Context) error
+}
+
+type DefaultAuthenticationCheck struct {
+	authType AuthType
+	client   *http.Client
+	url      string
+}
+
+func NewDefault(authType AuthType) AuthenticationCheck {
+	return &DefaultAuthenticationCheck{
+		authType: authType,
+		client:   http.DefaultClient,
+		url:      resource,
+	}
+}
+
+// AuthenticationCheck performs the authentication check running in the Pod.
+func (ac *DefaultAuthenticationCheck) Check(ctx context.Context) error {
+	var err error
+	switch ac.authType {
+	case Secret:
+		err = AuthenticationCheckForSecret(ctx)
+	case WorkloadIdentityGSA:
+		err = AuthenticationCheckForWorkloadIdentityGSA(ac.url, ac.client)
+	case WorkloadIdentity:
+		// Skip authentication check running in Pods which use new generation of Workload Identity.
+		return nil
+	default:
+		return fmt.Errorf("unknown auth type: %s", ac.authType)
+	}
+
+	if err != nil {
+		return writeTerminationLog(err, ac.authType)
+	}
+	return nil
+}
+
+// AuthenticationCheckForSecret performs the authentication check for Pod in secret mode.
+func AuthenticationCheckForSecret(ctx context.Context) error {
+	cred, err := google.FindDefaultCredentials(ctx, scope)
+	if err != nil {
+		return fmt.Errorf("error finding the default credential: %w", err)
+	}
+	s, err := cred.TokenSource.Token()
+	if err != nil {
+		return fmt.Errorf("error getting the token, probably due to the key stored in the Kubernetes Secret is expired or revoked: %w", err)
+	}
+	if !s.Valid() {
+		return errors.New("token is not valid")
+	}
+	return nil
+}
+
+// AuthenticationCheckForWorkloadIdentityGSA performs the authentication check for Pod in workload-identity-gsa mode.
+func AuthenticationCheckForWorkloadIdentityGSA(resource string, client *http.Client) error {
+	req, err := http.NewRequest(http.MethodGet, resource, nil)
+	if err != nil {
+		return fmt.Errorf("error setting up the http request: %w", err)
+	}
+	req.Header.Set("Metadata-Flavor", "Google")
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error getting the http response: %w", err)
+	}
+	defer resp.Body.Close()
+	// Check if we can successfully get the token.
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusBadRequest {
+		return errors.New("the Pod is not fully authenticated, " +
+			"probably due to corresponding k8s service account and google service account do not establish a correct relationship")
```

Comment:
So that we have some pointer if we need to debug, please add the response code into the error.

```suggestion
		return fmt.Errorf("the Pod is not fully authenticated, " +
			"probably due to corresponding k8s service account and google service account do not establish a correct relationship: %d", resp.StatusCode)
```